### PR TITLE
Implement settings channel for the Linux shell

### PR DESCRIPTION
**Pull request by robert-ancell**
#22435  Description

Implement settings channel for the Linux shell

## Related Issues

https://github.com/flutter/flutter/issues/65591

## Tests

Tests added to confirm settings are sent on engine startup.

Manually tested by changing the following in gnome-control-center:
- Accessibility > Large Text (sets text scale to 1.25)
- Date & Time > Time Format (sets clock to 12/24hr).
- Appearance > Window colors (sets brightness to dark. Ubuntu extension, use gnome-tweak-tool for other systems).

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by robert-ancell** [comment]
Due to a bug in the Flutter snap if you try the tests the initial values never change. The workaround is to run the binary directly, and the fix will be in the snap.

File: shell/platform/linux/fl_settings_plugin.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+G_DECLARE_FINAL_TYPE(FlSettingsPlugin,
+                     fl_settings_plugin,
+                     FL,
+                     SETTINGS_PLUGIN,
+                     GObject);
+
+/**
+ * FlSettingsPlugin:
+ *
+ * #FlSettingsPlugin is a plugin that implements FIXME.
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_settings_plugin.h:21]: I could really use a plugin that can do that 😀

File: shell/platform/linux/fl_view.cc [hunk 2]
```diff
@@ -239,6 +243,8 @@ static void fl_view_realize(GtkWidget* widget) {
     g_warning("Failed to start Flutter engine: %s", error->message);
     return;
   }
+
+  fl_settings_plugin_start(self->settings_plugin);
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_view.cc:247]: Is there a reason to put this on the view rather than the engine?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Please add unit testing that the message is being sent on startup; see https://github.com/flutter/engine/pull/22323 for an example.

File: shell/platform/linux/fl_settings_plugin.h (hunk 1 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_settings_plugin.h:21]: It's on my TODO list.

File: shell/platform/linux/fl_view.cc (hunk 2 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_view.cc:247]: I couldn't decide where to put it, so I put it with the other plugins for consistency. It doesn't requite anything specific from the view, but then neither does the platform plugin.

I think if we have some of them attached to the engine and some attached to the view then we need to do some renaming to make it clearer what is different about them. The name plugin is already confusing as they really aren't plugins at all.

Comment (stuartmorgan) [reply shell/platform/linux/fl_view.cc:247]: Yes, that's something that propagated from mobile, but in retrospect I regret following that naming. We could consider a mass change at some point.

macOS has all its "plugins" on the view class for legacy reasons; in practice only the ones related to the view (e.g., text input) should live there. A headless engine should still have the right time format, for instance.

**Comment by chinmaygarde** [comment]
@robert-ancell Any progress on the unit-tests requested by @stuartmorgan ?

**Comment by robert-ancell** [comment]
I'm currently working on a branch to add some base engine tests then will update this one with the new test for the settings channel.

File: shell/platform/linux/fl_view.cc (hunk 2 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_view.cc]: I've moved this plugin to the engine.

**Comment by robert-ancell** [comment]
This includes the change in https://github.com/flutter/engine/pull/22900, which should probably land before this one.

**Review by stuartmorgan — APPROVED** [review]
LGTM

**Pull request merged**